zstd-safe/src/seekable.rs

```diff
@@ -0,0 +1,721 @@
+//! worth of extra data, instead of the entire archive.
+
+use core::{marker::PhantomData, ptr::NonNull};
+
+use crate::{
+    parse_code, ptr_mut, ptr_void, CompressionLevel, InBuffer, OutBuffer,
+    SafeResult, WriteBuf, SEEKABLE_FRAMEINDEX_TOOLARGE,
+};
+
+/// Indicates that the passed frame index is too large.
+///
+/// This happens when `frame_index > num_frames()`.
+#[derive(Debug, PartialEq)]
+pub struct FrameIndexTooLargeError;
+
+impl core::fmt::Display for FrameIndexTooLargeError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str("Frame index too large")
+    }
+}
+
+/// Required to tracking streaming operation.
+///
+/// Streaming objects are reusable to avoid allocation and deallocation,
+/// to start a new compression operation call `init()`.
+pub struct SeekableCStream(NonNull<zstd_sys::ZSTD_seekable_CStream>);
+
+unsafe impl Send for SeekableCStream {}
+unsafe impl Sync for SeekableCStream {}
+
+impl Default for SeekableCStream {
+    fn default() -> Self {
+        SeekableCStream::create()
+    }
+}
+
+impl SeekableCStream {
+    /// Tries to create a new `SeekableCStream`.
+    ///
+    /// Returns `None` if zstd returns a NULL pointer - may happen if allocation fails.
+    pub fn try_create() -> Option<Self> {
+        // Safety: Just FFI
+        Some(SeekableCStream(NonNull::new(unsafe {
+            zstd_sys::ZSTD_seekable_createCStream()
+        })?))
+    }
+
+    /// Creates a new `SeekableCStream`.
+    ///
+    /// # Panics
+    ///
+    /// If zstd returns a NULL pointer.
+    pub fn create() -> Self {
+        Self::try_create()
+            .expect("zstd returned null pointer when creating new seekable compression stream")
+    }
+
+    /// Wraps the `ZSTD_seekable_initCStream()` function.
+    ///
+    /// Call this to initialize a `SeekableCStream` object for a new compression operation.
+    /// - `max_frame_size` indicates the size at which to automatically start a new seekable
+    /// frame. `max_frame_size == 0` implies the default maximum size. Smaller frame sizes allow
+    /// faster decompression of small segments, since retrieving a single byte requires
+    /// decompression of the full frame where the byte belongs. In general, size the frames
+    /// to roughly correspond to the access granularity (when it's known). But small sizes
+    /// also reduce compression ratio. Avoid really tiny frame sizes (< 1 KB), that would
+    /// hurt compression ratio considerably.
+    /// - `checksum_flag` indicates whether or not the seek table should include frame
+    /// checksums on the uncompressed data for verification.
+    ///
+    /// Returns a size hint for input to provide for compression, or an error code.
+    pub fn init(
+        &mut self,
+        compression_level: CompressionLevel,
+        checksum_flag: bool,
+        max_frame_size: u32,
+    ) -> SafeResult {
+        let c_checksum_flag = if checksum_flag { 1 } else { 0 };
```

Comment:
You should also be able to do `checksum_flag as u8`.

Reply:
Works with `checksum_flag as i32`, thanks for the hint